server/src/internalClusterTest/java/org/elasticsearch/action/bulk/BulkAfterWriteFsyncFailureIT.java

```diff
@@ -0,0 +1,102 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
+
+public class BulkAfterWriteFsyncFailureIT extends ESSingleNodeTestCase {
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return List.of(FailingFsyncEnginePlugin.class);
+    }
+
+    public static class FailingFsyncEnginePlugin extends Plugin implements EnginePlugin {
+        static final AtomicBoolean simulateFsyncFailure = new AtomicBoolean(false);
+
+        @Override
+        public Optional<EngineFactory> getEngineFactory(IndexSettings indexSettings) {
+            return Optional.of(config -> new InternalEngine(config) {
+                @Override
+                protected ChannelFactory getTranslogChannelFactory() {
+                    return (file, openOption) -> new FilterFileChannel(FileChannel.open(file, openOption)) {
+                        @Override
+                        public void force(boolean metaData) throws IOException {
+                            if (simulateFsyncFailure.compareAndSet(true, false)) {
+                                throw new IOException("Simulated fsync failure");
+                            } else {
+                                super.force(metaData);
+                            }
+                        }
+                    };
+                }
+            });
+        }
+    }
+
+    public void testFsyncFailureDoesNotAdvanceLocalCheckpoints() {
+        String indexName = randomIdentifier();
+        client().admin()
+            .indices()
+            .prepareCreate(indexName)
+            .setSettings(
+                Settings.builder()
+                    .put(INDEX_REFRESH_INTERVAL_SETTING.getKey(), -1)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                    .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                    .build()
+            )
+            .get();
+        ensureGreen(indexName);
+
+        FailingFsyncEnginePlugin.simulateFsyncFailure.set(true);
+        var localCheckpointBeforeBulk = getLocalCheckpointForShard(indexName, 0);
+        var bulkResponse = client().prepareBulk().add(prepareIndex(indexName).setId("1").setSource("key", "foo", "val", 10)).get();
+        assertTrue(bulkResponse.hasFailures());
+        var localCheckpointAfterFailedBulk = getLocalCheckpointForShard(indexName, 0);
+        // fsync for the translog failed, hence the checkpoint doesn't advance
+        assertThat(localCheckpointBeforeBulk, equalTo(localCheckpointAfterFailedBulk));
+
+        // Since background refreshes are disabled, the shard is considered green until the next operation is appended into the translog
+        ensureGreen(indexName);
+
+        var bulkResponse2 = client().prepareBulk().add(prepareIndex(indexName).setId("2").setSource("key", "bar", "val", 20)).get();
+        assertFalse(bulkResponse2.hasFailures());
```

Comment:
Would we not expect this to also fail due to the tragic failure above?

Reply:
The TransportReplicationAction will retry on `AlreadyClosedException` https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/server/src/main/java/org/elasticsearch/action/support/TransportActions.java#L28

Reply:
oh, can we add a comment about that to the test? Seems the test concerns the engine level, so this retry is not obvious when reading it.